collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java

```diff
@@ -164,6 +173,66 @@ public void onDeleteRepeatClicked() {
         });
     }
 
+    private void handleInstanceEditResult(InstanceEditResult result) {
+        Instance instance = result.getInstance();
+
+        if (result instanceof InstanceEditResult.EditCompleted) {
+            openEditedInstance(instance.getDbId());
+        } else if (result instanceof InstanceEditResult.EditBlockedByNewerExistingEdit) {
+            if (InstanceExtKt.isDraft(instance)) {
+                showOpenDraftEditDialog(instance);
+            } else {
+                showOpenFinalizedEditDialog(instance);
+            }
+        }
+    }
+
+    private void openEditedInstance(long dbId) {
+        Intent intent = FormFillingIntentFactory.editFinalizedFormIntent(requireContext(), currentProjectId, dbId);
+        startActivity(intent);
+        requireActivity().getOnBackPressedDispatcher().onBackPressed();
+    }
+
+    private void showOpenDraftEditDialog(Instance instance) {
+        String dialogMessage = new SimpleDateFormat(
+                getString(org.odk.collect.strings.R.string.newer_draft_edit_found_dialog_message),
+                Locale.getDefault()
+        ).format(instance.getLastStatusChangeDate());
+
+        new MaterialAlertDialogBuilder(requireContext())
```

Comment:
It'd definitely help us down the line to convert `FormHierarchyFragmentHostActivity` to using a nav graph and then have the dialogs here just be `DialogFragment` implementations that are navigated to with the `NavigationController`. This is something we'll want to do anyway if we want form entry/hierarchy to be one Activity using a nav graph. Is that something you'd be up for trying out as a follow-up?